src/Featurizers/Components/NormUpdaters.h

```diff
@@ -1,298 +1,381 @@
+
+    // flags to check if update function is called
+    // an error will be thrown if no valid input is input before commit
+    bool        _has_valid_input;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+    void update_impl(T input, std::true_type /*is_nullable*/);
+    void update_impl(T input, std::false_type /*is_nullable*/);
+
+    template <typename U>
+    void update_impl(U input);
+};
+
+} // Updaters
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  Updaters::L1NormUpdater
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Updaters::L1NormUpdater<T>::L1NormUpdater() {
+    reset();
+}
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::reset() {
+    _l1_norm = 0;
+    _has_valid_input = false;
+}
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::update(T input) {
+    update_impl(input, std::integral_constant<bool, Traits<T>::IsNullableType>());
+}
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::update_impl(T input, std::true_type /*is_nullable*/) {
+    if (Traits<T>::IsNull(input)) {
+        return;
+    }
+
+    update_impl(Traits<T>::GetNullableValue(input));
+}
+
+
+template <typename T>
+void Updaters::L1NormUpdater<T>::update_impl(T input, std::false_type /*is_nullable*/) {
+    update_impl(input);
+}
+
+template <typename T>
+template <typename U>
+void Updaters::L1NormUpdater<T>::update_impl(U input) {
+    if (!_has_valid_input)
+        _has_valid_input = true;
+
+    long double diff = std::abs(static_cast<long double>(input));
+
+    // check if diff is too small comparing to l1_norm
+    if ((_l1_norm + diff == _l1_norm) && (diff != 0)) {
+        throw std::runtime_error("In l1 norm calculation, input is so small comparing to l1_norm that l1_norm is the same after long double addition!");
+    }
+    _l1_norm += diff;
+
+    // check if exceeds bound
+    // only check for l2_norm and no checks for l1_norm is because if l1_norm is out of upper bound, l2_norm is definitely out of bound
+    if (std::isinf(_l1_norm)) {
+        throw std::runtime_error("Overflows occured during calculating l1_norm! Check your data!");
+    }
+}
+
+template <typename T>
+long double Updaters::L1NormUpdater<T>::commit(void) {
+    assert(_l1_norm >= 0);
```

Comment:
>assert(_l1_norm >= 0); [](start = 4, length = 22)

Nit: move this below the if check. We shouldn't look at _l1_norm until we know that it is valid.